### PR TITLE
fix replica set hot loop

### PR DESCRIPTION
**Pull request by m1093782566**
**What this PR does / why we need it**:

Fix replicas set hot loop.

Related issue: #30629

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33092)

### CODE REVIEW

**Comment by m1093782566** [comment]
@deads2k @derekwaynecarr PTAL.

Thanks!

File: pkg/controller/replicaset/replica_set.go
Comment (mikedanese) [inline pkg/controller/replicaset/replica_set.go]:
```go
wait.Forever(rsc.processNextWorkItem, 0)
```
or
```go
for {
    rsc.processNextWorkItem()
}
```
might read better

File: pkg/controller/replicaset/replica_set.go
Comment (mikedanese) [inline pkg/controller/replicaset/replica_set.go]: use '%q' for key

File: pkg/controller/replicaset/replica_set.go
Comment (mikedanese) [inline pkg/controller/replicaset/replica_set.go]: This looks not formated?

File: pkg/controller/replicaset/replica_set.go
Comment (mikedanese) [inline pkg/controller/replicaset/replica_set.go]: please delete commented out code

**Review by mikedanese — CHANGES_REQUESTED** [review]
Thanks, this looks good. Some small comments.

File: pkg/controller/replicaset/replica_set.go
Comment (m1093782566) [inline pkg/controller/replicaset/replica_set.go]: Hmm...

```
for rsc.processNextWorkItem() {
}
```

is not a dead loop. Since `processNextWorkItem()` may return false and will end the loop.

File: pkg/controller/replicaset/replica_set.go
Comment (m1093782566) [inline pkg/controller/replicaset/replica_set.go]: Okay, will fix.

File: pkg/controller/replicaset/replica_set.go
Comment (m1093782566) [inline pkg/controller/replicaset/replica_set.go]: Okay, will go fmt code

File: pkg/controller/replicaset/replica_set.go
Comment (m1093782566) [inline pkg/controller/replicaset/replica_set.go]: Okay, will do.

File: pkg/controller/replicaset/replica_set.go
Comment (mikedanese) [inline pkg/controller/replicaset/replica_set.go]: ah ok, thanks for pointing that out.

**Comment by m1093782566** [comment]
@mikedanese comments addressed, PTAL.

Thanks!

File: pkg/controller/replicaset/replica_set.go
Comment (kargakis) [inline pkg/controller/replicaset/replica_set.go]: This error will be logged anyway in `processNextWorkItem`

File: pkg/controller/replicaset/replica_set.go
Comment (kargakis) [inline pkg/controller/replicaset/replica_set.go]: It's probably worth mentioning that we will requeue the replica set now in case of an error while creating / deleting pods. Seems that we didn't do it prior to this change.

File: pkg/controller/replicaset/replica_set.go
Comment (m1093782566) [inline pkg/controller/replicaset/replica_set.go]: Ah, yes, will fix.

File: pkg/controller/replicaset/replica_set.go
Comment (m1093782566) [inline pkg/controller/replicaset/replica_set.go]: Hmm, I agree with you and I will add some lines comments.

Comment (m1093782566) [inline pkg/controller/replicaset/replica_set.go]: Fixed.

**Comment by m1093782566** [comment]
@kargakis comments addressed. PTAL.

**Comment by kargakis** [comment]
LGTM, pending a review from @mikedanese

**Review by mikedanese — APPROVED** [review]

**Comment by m1093782566** [comment]
pkg/controller/replicaset/replica_set_test.go:1157: possible formatting directive in Fatal call

exit status 1

Makefile:99: recipe for target 'verify' failed

make: *** [verify] Error 1

**Comment by m1093782566** [comment]
go run hack/e2e.go -v -test --test_args='--ginkgo.focus=Up$'

error running up: exit status 1from junit_runner.xml

**Comment by bprashanth** [comment]
please verify that replication controller also doesn't suffer from the same issue (I haven't opened the pr, just noticed that files touched didn't include replication_conroller.go)

**Comment by m1093782566** [comment]
> please verify that replication controller also doesn't suffer from the same issue

I think @deads2k has fixed replication controller hot loop?

**Comment by kargakis** [comment]
yes

**Comment by m1093782566** [comment]
kubernetes-pull-build-test-gci-kubemark-e2e-gce #20

Recent runs
PR	m1093782566: fix replica set hot loop
Result	FAILURE
Started	2016-09-24 08:57 CST
Elapsed	14m30s
Version	v1.5.0-alpha.0.1316+971567d5f6a819
Builder	agent-pr-24

**Comment by m1093782566** [comment]
CI failed by flake issue #33459

Can anyone please ping rebot?

**Comment by bprashanth** [comment]
oh 
```
WARNING: You are creating a legacy network. Using --mode=legacy will be required in future releases.
ERROR: (gcloud.compute.networks.create) Some requests did not succeed:
 - Quota 'NETWORKS' exceeded. Limit: 5.0
```

I guess we're leaking networks? @kubernetes/goog-testing (networks is a rare leak, I'll clean them up but the source is unknown)

**Comment by bprashanth** [comment]
@k8s-bot test this github issue: #IGNORE

**Comment by bprashanth** [comment]
Ah, this failure looks more familiar: https://github.com/kubernetes/kubernetes/issues/33617
I think today's just a bad day for flakes. Going to wait a bit before kicking it. If someone doesn't get to debuggin the issue, I probably will at some point, and then we can restart the test.

**Comment by kargakis** [comment]
@bprashanth can we move forward with merging the PR since the change seems unrelated to the flake? I would like to have this merged and add Conditions on top for both controllers.

**Comment by bprashanth** [comment]
@k8s-bot test this github issue: https://github.com/kubernetes/kubernetes/issues/33617

**Comment by kargakis** [comment]
@k8s-bot test this issue: https://github.com/kubernetes/kubernetes/issues/33388

**Comment by kargakis** [comment]
@k8s-bot gci test this issue: #33388

**Comment by kargakis** [comment]
How can we restart the Jenkins GCI Kubemark GCE e2e suite?

File: pkg/controller/replicaset/replica_set.go
Comment (kargakis) [inline pkg/controller/replicaset/replica_set.go]: The RC controller is not aggregating errors but instead it will return them separately: there are only two of them anyway, one error representing a replica create/delete failure and one about a failed RC status update. In the first case it will retry forever albeit being rate-limited and in the second case it will only retry once more. There is no reason for not having both controllers being consistent as far as error handling is concerned.  I am not sure though if the RC controller is 100% correct.

Comment (kargakis) [inline pkg/controller/replicaset/replica_set.go]: Actually there is no difference on a second look I had. Still I would prefer to follow the same pattern as the RC controller and not aggregate here.

Comment (m1093782566) [inline pkg/controller/replicaset/replica_set.go]: Thanks for your advice, I will have a deep look at RC controller.

Comment (kargakis) [inline pkg/controller/replicaset/replica_set.go]: It should be an easy fix: you basically don't aggregate here and return just the manageReplicas error. In the if clause above you return the status update error instead of appending to the list.

**Comment by m1093782566** [comment]
@kargakis 

Sorry for delay, comments addressed. PTAL.

**Pull request merged**
